microbench/src/main/java/io/netty/handler/codec/http/DecodeHexBenchmark.java

```diff
@@ -0,0 +1,136 @@
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.microbench.util.AbstractMicrobenchmark;
+import io.netty.util.internal.StringUtil;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+@Warmup(iterations = 5, time = 1)
+@Measurement(iterations = 5, time = 1)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+public class DecodeHexBenchmark extends AbstractMicrobenchmark {
+
+    @Param({ "abcdefABCDEF0123456789g", "abcdefABCDEF0123456789" })
+    private String hex;
+    private char[] hexDigits;
+
+    @Setup
+    public void init() {
+        hexDigits = hex.toCharArray();
+    }
+
+    @Benchmark
+    public long hexDigits() {
+        long v = 0;
+        final char[] hexDigits = this.hexDigits;
+        for (int i = 0, size = hexDigits.length; i < size; i++) {
+            v += StringUtil.decodeHexNibble(hexDigits[i]);
+        }
+        return v;
+    }
+
+    @Benchmark
+    public long hexDigitsBranchlessNoChecks() {
+        long v = 0;
+        final char[] hexDigits = this.hexDigits;
+        for (int i = 0, size = hexDigits.length; i < size; i++) {
+            v += branchlessNoCheck(hexDigits[i]);
+        }
+        return v;
+    }
+
+    @Benchmark
+    public long hexDigitsUsingArrayWithChecks() {
+        long v = 0;
+        final char[] hexDigits = this.hexDigits;
+        for (int i = 0, size = hexDigits.length; i < size; i++) {
+            v += lookupWithCheck(hexDigits[i]);
+        }
+        return v;
+    }
+
+    @Benchmark
+    public long hexDigitsUsingArrayWithEx() {
+        long v = 0;
+        final char[] hexDigits = this.hexDigits;
+        for (int i = 0, size = hexDigits.length; i < size; i++) {
+            v += lookupWithEx(hexDigits[i]);
+        }
+        return v;
+    }
+
+    private static final int[] HEX;
+
+    static {
+        HEX = new int['f' + 1];
+        Arrays.fill(HEX, -1);
+        HEX['0'] = 0;
+        HEX['1'] = 1;
+        HEX['2'] = 2;
+        HEX['3'] = 3;
+        HEX['4'] = 4;
+        HEX['5'] = 5;
+        HEX['6'] = 6;
+        HEX['7'] = 7;
+        HEX['8'] = 8;
+        HEX['9'] = 9;
+        HEX['a'] = 10;
+        HEX['b'] = 11;
+        HEX['c'] = 12;
+        HEX['d'] = 13;
+        HEX['e'] = 14;
+        HEX['f'] = 15;
+        HEX['A'] = 10;
+        HEX['B'] = 11;
+        HEX['C'] = 12;
+        HEX['D'] = 13;
+        HEX['E'] = 14;
+        HEX['F'] = 15;
+    }
+
+    public static int lookupWithCheck(final char c) {
+        final int index = c;
+        if (index < 0 || index >= HEX.length) {
+            return -1;
+        }
+        return HEX[c];
```

Comment:
Oh realized this might be why - what happens if you change `c` to `index`?

Reply:
Uh eagle eyes! I have missed that one!!I need to arrive at the office but that could be the reason

Reply:
Is `< 0` still necessary? Is JVM smart enough to guess `index` is always >= 0?

Reply:
Would be surprised - at the bytecode level it's all `int` so don't think the JIT would have any way to know this. If it does it's true magic :)

Reply:
No idea :) I've tried to mimic the same pattern of the check that the JVM would have performed. In few minutes I'll arrive at the office and will run again the bench checking the ASM because there are a couple of thing not 100% clear to me..anyway I see that the lookup table is delivering and interesting speedup, yet to see how it will impact overall...already done some benchs on https://github.com/line/armeria/pull/2307? Just curious ;)